### ISSUE TITLE
[Issue]: bf16 MLA decode on gfx942 reads the wrong KV past 4 GiB from ptr_KV

### ISSUE THREAD

**Issue by jamesETsmith** [ROCm/aiter#5826]
### Problem Description

On gfx942, the bf16 MLA decode ASM kernels address the KV cache with a 32-bit byte offset from `ptr_KV`. Once a row sits at or past `2**32` bytes from that pointer, the kernel reads the wrong memory and returns a wrong attention output. There is no fault, no NaN, and no host warning.

A row is 576 bf16 elements (512 latent + 64 RoPE), which is 1152 bytes, so the cliff is exact:

```
2**32 // 1152 == (2**31 - 1) // 576 == 3,728,270
```

Row 3,728,269 is fully correct. Row 3,728,270 straddles 4 GiB and comes back with only the first 120 of 512 latent outputs correct. Row 3,728,271 is wrong from the first element. A truncated row index cannot do that: the row ids fit in int32, and the page stride (1152) fits in int32. The product `index * stride` is what wraps, as a byte offset added to `ptr_KV`.

The same bytes are correct when the base pointer moves so the 32-bit offset stays small. Triton MLA (`aiter.ops.triton.attention.mla.mla_decode_fwd`) stays correct on the overflowing placement. `aiter.mla._check_mla_kv_int32_span` is not defined at this commit, so nothing on the host masked the kernel.

This shows up with a pool-global page index. One request can be far shorter than 3.7M tokens and still index a row past 4 GiB in a shared KV slab.

## Kernels that wrap

All four paths below were wrong past the cliff and correct below it. `max_seqlen_q=1` on the persistent path is dispatched as `qSeqLen=4`.

| path | code object | past 4 GiB |
|---|---|---|
| persistent, 16 heads | `hsa/gfx942/mla/mla_a16w16_qh16_m16x4_n16x1_coex0_mask1_ps.co` | wraps |
| stage-1, 16 heads (no `work_meta_data`) | `hsa/gfx942/mla/mla_dec_stage1_bf16_a16w16_subQ16_mqa16.co` | wraps |
| persistent, 128 heads | same qh16 persistent `.co` (no separate blob loaded) | wraps |
| stage-1, 128 heads | `hsa/gfx942/mla/mla_dec_stage1_bf16_a16w16_subQ128_mqa128.co` | wraps |

### Operating System

24.04.3 LTS (Noble Numbat)

### CPU

Intel(R) Xeon(R) Platinum 8480+

### GPU

AMD Instinct MI325X

### ROCm Version

ROCm 10.0

### Installation Method

Other (describe in Additional Information)

### Installed ROCm Packages / Versions

<details>
<summary>Installed ROCm packages / versions</summary>

Used docker rocm/pytorch:rocm10.0_ubuntu24.04_py3.12_pytorch_release_2.12.0

</details>

### Steps to Reproduce

<details>
<summary>Repro script</summary>

```python
"""Repro: persistent-scheduling ASM MLA decode wraps at 32-bit KV offsets.

python op_tests/repro_mla_int32_kv_offset.py

> expected output
  low            offsets=[0, 18,873,792] asm_err=0.00103 (OK) triton_err=0.00056 (OK)
  edge_safe      offsets=[2,128,609,728, 2,147,483,520] asm_err=0.00452 (OK) triton_err=0.00056 (OK)
  edge_overflow  offsets=[2,147,484,096, 2,166,357,888] asm_err=0.21891 (WRAPPED) triton_err=0.00056 (OK)

The same query and the same KV rows are placed at three offsets in one
paged pool:

  low            -- element offsets near zero
  edge_safe      -- max element offset just below 2**31
  edge_overflow  -- min element offset just above 2**31

For each placement we compare:

  * torch reference (absorbed MLA: softmax(q k^T) @ k[:, :512])
  * persistent-scheduling ASM decode (mla_a16w16_*_ps / mla_decode_fwd with
    work_meta_data)
  * Triton MLA decode (64-bit-safe addressing control)

Observed on gfx942 (MI325X), bf16, D=576 (512 latent + 64 rope), H=16:

  low            : ASM matches reference
  edge_safe      : ASM matches reference (max element offset < 2**31)
  edge_overflow  : ASM returns wrapped/garbage attention output while the
                   Triton kernel and the torch reference still match -- the KV
                   data is intact; only the ASM kernel's 32-bit element
                   addressing wraps.

Usage (needs a GPU with ~4.1 GiB free; default pool is ~3.76M rows x 576):

  PYTHONPATH=. python op_tests/repro_mla_int32_kv_offset.py
  PYTHONPATH=. python op_tests/repro_mla_int32_kv_offset.py --batch 1 --seq-len 1024
"""

import argparse

import torch

D = 576  # kv_lora_rank (512) + qk_rope_head_dim (64)
V = 512  # kv_lora_rank
INT32_LIMIT = 2**31
# Last row whose element offset still fits int32: row * D <= INT32_LIMIT - 1.
LIMIT_ROW = (INT32_LIMIT - 1) // D


def make_metadata(batch, heads, qo_indptr, kv_indptr, last_page_lens, device):
    from aiter import dtypes, get_mla_metadata_info_v1, get_mla_metadata_v1

    sizes = get_mla_metadata_info_v1(
        batch, 1, heads, dtypes.bf16, dtypes.bf16, is_sparse=False, fast_mode=True
    )
    bufs = []
    for size, dtype in sizes:
        shape = size if isinstance(size, tuple) else int(size)
        bufs.append(torch.empty(shape, dtype=dtype, device=device))
    get_mla_metadata_v1(
        qo_indptr,
        kv_indptr,
        last_page_lens,
        heads,
        1,
        True,
        bufs[0],
        bufs[2],
        bufs[1],
        bufs[3],
        bufs[4],
        bufs[5],
        page_size=1,
        kv_granularity=16,
        max_seqlen_qo=1,
        uni_seqlen_qo=1,
        fast_mode=True,
        dtype_q=dtypes.bf16,
        dtype_kv=dtypes.bf16,
    )
    return bufs


def reference(q, kv_rows, scale):
    # Absorbed MLA: per head, out = softmax(scale * q k^T) @ k[:, :V].
    b, h, _ = q.shape
    out = torch.empty(b, h, V, dtype=torch.float32, device=q.device)
    k = kv_rows.float()
    for bi in range(b):
        scores = scale * (q[bi].float() @ k[bi].T)
        probs = torch.softmax(scores, dim=-1)
        out[bi] = probs @ k[bi][:, :V]
    return out


def run_case(name, pool, start_rows, q, kv_source, args):
    from aiter.mla import mla_decode_fwd

    device = q.device
    b, seq = args.batch, args.seq_len
    scale = 1.0 / (D**0.5)

    # Place identical KV rows at the requested offsets.
    for bi in range(b):
        pool[start_rows[bi] : start_rows[bi] + seq] = kv_source[bi].view(seq, 1, 1, D)

    kv_indices = torch.cat(
        [torch.arange(s, s + seq, dtype=torch.int32, device=device) for s in start_rows]
    )
    qo_indptr = torch.arange(0, b + 1, dtype=torch.int32, device=device)
    kv_indptr = torch.arange(0, b + 1, dtype=torch.int32, device=device) * seq
    last_page_lens = torch.ones(b, dtype=torch.int32, device=device)
    meta = make_metadata(b, args.heads, qo_indptr, kv_indptr, last_page_lens, device)
    out_asm = torch.empty(b, args.heads, V, dtype=torch.bfloat16, device=device)
    mla_decode_fwd(
        q,
        pool.view(-1, 1, 1, D),
        out_asm,
        qo_indptr,
        kv_indptr,
        kv_indices,
        last_page_lens,
        1,
        sm_scale=scale,
        work_meta_data=meta[0],
        work_indptr=meta[1],
        work_info_set=meta[2],
        reduce_indptr=meta[3],
        reduce_final_map=meta[4],
        reduce_partial_map=meta[5],
    )

    # 64-bit-safe Triton control on the same rows.
    from aiter.ops.triton.attention.mla import mla_decode_fwd as triton_mla_decode_fwd

    out_triton = torch.empty(b, args.heads, V, dtype=torch.bfloat16, device=device)
    block_tables = kv_indices.view(b, seq)
    triton_mla_decode_fwd(
        q=q,
        kv_buffer=pool.view(-1, 1, 1, D),
        out=out_triton,
        cu_seqlens_q=qo_indptr,
        seqused_k=torch.full((b,), seq, dtype=torch.int32, device=device),
        max_seqlen_kv=seq,
        block_tables=block_tables,
        softmax_scale=scale,
        kv_lora_rank=V,
        qk_rope_head_dim=D - V,
        causal=True,
        q_descale=None,
        kv_descale=None,
    )
    torch.cuda.synchronize()

    kv_rows = torch.stack([pool[s : s + seq].view(seq, D) for s in start_rows])
    ref = reference(q, kv_rows, scale)
    asm_err = (out_asm.float() - ref).abs().max().item()
    tri_err = (out_triton.float() - ref).abs().max().item()
    max_off = int(max(start_rows) + seq - 1) * D
    min_off = int(min(start_rows)) * D

    asm_ok = torch.allclose(out_asm.float(), ref, atol=2e-2, rtol=2e-2)
    tri_ok = torch.allclose(out_triton.float(), ref, atol=2e-2, rtol=2e-2)
    print(
        f"{name:14} offsets=[{min_off:,}, {max_off:,}] "
        f"asm_err={asm_err:.5f} ({'OK' if asm_ok else 'WRAPPED'}) "
        f"triton_err={tri_err:.5f} ({'OK' if tri_ok else 'BAD'})",
        flush=True,
    )
    # Reset placement so cases are independent.
    for bi in range(b):
        pool[start_rows[bi] : start_rows[bi] + seq].zero_()
    return asm_ok, tri_ok


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch", type=int, default=8)
    parser.add_argument("--seq-len", type=int, default=4096)
    parser.add_argument("--heads", type=int, default=16)
    args = parser.parse_args()

    assert torch.cuda.is_available(), "this repro needs a GPU"
    device = torch.device("cuda")
    torch.manual_seed(42)

    rows_needed = args.batch * args.seq_len
    pool_rows = LIMIT_ROW + rows_needed + 8
    pool_bytes = pool_rows * D * 2
    free = torch.cuda.mem_get_info()[0]
    print(
        f"pool_rows={pool_rows:,} bytes={pool_bytes / 2**30:.2f} GiB "
        f"free={free / 2**30:.2f} GiB limit_row={LIMIT_ROW:,} "
        f"(row*{D} vs 2**31)",
        flush=True,
    )
    if pool_bytes + (1 << 30) > free:
        raise SystemExit(
            "not enough GPU memory; rerun with --batch 1 --seq-len 1024 "
            f"(needs ~{(LIMIT_ROW + 1024) * D * 2 / 2**30:.2f} GiB)"
        )

    pool = torch.zeros(pool_rows, 1, 1, D, dtype=torch.bfloat16, device=device)
    q = torch.randn(args.batch, args.heads, D, dtype=torch.bfloat16, device=device)
    kv_source = torch.randn(args.batch, args.seq_len, D, dtype=torch.bfloat16, device=device)

    # This repro intentionally bypasses the host-side span guard (when present)
    # so the underlying kernel behavior stays observable.
    import aiter.mla as mla_mod

    if hasattr(mla_mod, "_check_mla_kv_int32_span"):
        mla_mod._check_mla_kv_int32_span = lambda *a, **k: None

    seq = args.seq_len
    results = []
    results.append(
        run_case("low", pool, [bi * seq for bi in range(args.batch)], q, kv_source, args)
    )
    safe_start = LIMIT_ROW - rows_needed + 1
    results.append(
        run_case(
            "edge_safe",
            pool,
            [safe_start + bi * seq for bi in range(args.batch)],
            q,
            kv_source,
            args,
        )
    )
    over_start = LIMIT_ROW + 1
    results.append(
        run_case(
            "edge_overflow",
            pool,
            [over_start + bi * seq for bi in range(args.batch)],
            q,
            kv_source,
            args,
        )
    )
    print(
        "summary: asm should be OK/OK/WRAPPED and triton OK/OK/OK; "
        "edge_overflow WRAPPED means the persistent ASM kernel's 32-bit "
        "element addressing wrapped while the KV data is intact.",
        flush=True,
    )


if __name__ == "__main__":
    main()

```
</details>

### (Optional for Linux users) Output of rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

### Related Issues
- #5765 same issue but for gfx950
